File: weather/dart/weather.dart
#import('dart:html');
#import('dart:json');

void main() {
  query('#formulary').on.submit.add(newQuery);
  query('#form_login').on.submit.add(login);
  query('#logout').on.click.add(logout);
}

void login(event){
  XMLHttpRequest req = new XMLHttpRequest();
  req.open("POST", "http://192.168.1.63:8000/weather/login/", true);
  req.withCredentials = true;
  
  InputElement username = query('#user_id');
  var user = username.value;
  InputElement password = query('#pass_id');
  var pass = password.value;
  
  // Add an event handler to call the onSuccess function when the POST call is successful
  req.on.readyStateChange.add((Event e) {
    if (req.readyState == XMLHttpRequest.DONE && (req.status == 200)) {
      DivElement div_loggin = query('#log_form');
      div_loggin.style.display = 'none';
      
      DivElement logged_user = query('#logged_user');
      logged_user.style.display = '';
      logged_user.text = user;
      
      DivElement div_loggout_button = query('#logout_button');
      div_loggout_button.style.display = '';
    } else {
      print(req.status);
      print("ERROR DE LOGIN!!!!");
    }
  });

  var data = JSON.stringify({
    'username' : user,
    'password' : pass
  });
  
  req.send(data);
  event.preventDefault();
}

void logout(event) {
  XMLHttpRequest req = new XMLHttpRequest();
  req.open("POST", "http://192.168.1.63:8000/weather/logout/", true);
  req.withCredentials = true;
  req.send();
  
  DivElement div_loggin = query('#log_form');
  div_loggin.style.display = '';
  
  DivElement logged_user = query('#logged_user');
  logged_user.style.display = 'none';
  
  DivElement div_loggout_button = query('#logout_button');
  div_loggout_button.style.display = 'none';
  
  DivElement div_query = query('#content');
  div_query.style.display = 'none';
}

void newQuery(event) {
  // Get the city name from the html form
  InputElement input = query("#city");
  var city = input.value;
  
  // Create and open a new request
  XMLHttpRequest req = new XMLHttpRequest();
  req.open("POST", "http://192.168.1.63:8000/weather/querys/", true);
  req.withCredentials = true;
  
  // Add an event handler to call the onSuccess function when the POST call is successful
  req.on.readyStateChange.add((Event e) {
    if (req.readyState == XMLHttpRequest.DONE && (req.status == 200)) {
      var j = JSON.parse(req.responseText);
      if (j.containsKey("data") && j["data"].containsKey("error"))
        print('HELL YEAH');
      else
        onSuccess(req);
    }
  });
  
  // Set the content type of the call and send the data
  req.setRequestHeader('Content-Type', 'application/json');
  req.send(JSON.stringify({'city': city}));

  event.preventDefault();
}

// Function to call after a successful POST call
void onSuccess(XMLHttpRequest req) {
  DivElement div = query('#content');
  div.style.display = '';
  showJson(req);
}

// Function that shows the desired message in the corresponding element
void showJson(XMLHttpRequest req) {
  var parsedJSON = JSON.parse(req.responseText)["result_json"];
  var currentConditions = parsedJSON["data"]["current_condition"][0];
  query('#observationTime').text = currentConditions["observation_time"];
  query('#description1').text = currentConditions["weatherDesc"][0]["value"];
  query('#tCelsius').text = '${currentConditions["temp_C"]}ºC';
  query('#tFarenheit').text = '${currentConditions["temp_F"]}ºF';
  query('#humidity').text = currentConditions["humidity"];
  query('#precipitation1').text = '${currentConditions["precipMM"]} MM';
  query('#cloudcover').text =  '${currentConditions["cloudcover"]}%';
  query('#visibility').text =  currentConditions["visibility"];
  query('#pressure').text =  currentConditions["pressure"];
  query('#windDirectionDegree1').text =  '${currentConditions["winddirDegree"]}º';
  query('#windDirection16Point1').text =  currentConditions["winddir16Point"];
  query('#windSpeedkmh1').text =  '${currentConditions["windspeedKmph"]}Kmph';
  query('#windSpeedMiles1').text =  '${currentConditions["windspeedMiles"]}Mph';
  ImageElement image = query('#img1');
  image.src = currentConditions["weatherIconUrl"][0]["value"];
  query('#city_name').text = parsedJSON["data"]["request"][0]["query"];
  var weatherToday = parsedJSON["data"]["weather"][0];
  query('#dateToday').text = weatherToday["date"];
  query('#description2').text = weatherToday["weatherDesc"][0]["value"];
  query('#minTCelsius1').text = '${weatherToday["tempMinC"]}ºC';
  query('#minTFarenheit1').text = '${weatherToday["tempMinF"]}ºF';
  query('#maxTCelsius1').text = '${weatherToday["tempMaxC"]}ºC';
  query('#maxTFarenheit1').text = '${weatherToday["tempMaxF"]}ºF';
  query('#precipitation2').text = weatherToday["precipMM"];
  query('#precipitation2').text = weatherToday["precipMM"];
  query('#windDirectionDegree2').text =  '${weatherToday["winddirDegree"]}º';
  query('#windDirection16Point2').text =  weatherToday["winddir16Point"];
  query('#windSpeedkmh2').text =  '${weatherToday["windspeedKmph"]}Kmph';
  query('#windSpeedMiles2').text =  '${weatherToday["windspeedMiles"]}Mph';
  ImageElement image2 = query('#img2');
  image2.src = weatherToday["weatherIconUrl"][0]["value"];
  
  var weatherTomorrow = parsedJSON["data"]["weather"][1];
  query('#dateTomorrow').text = weatherTomorrow["date"];
  query('#description3').text = weatherTomorrow["weatherDesc"][0]["value"];
  query('#minTCelsius2').text = '${weatherTomorrow["tempMinC"]}ºC';
  query('#minTFarenheit2').text = '${weatherTomorrow["tempMinF"]}ºF';
  query('#maxTCelsius2').text = '${weatherTomorrow["tempMaxC"]}ºC';
  query('#maxTFarenheit2').text = '${weatherTomorrow["tempMaxF"]}ºF';
  query('#precipitation3').text = weatherTomorrow["precipMM"];
  query('#precipitation3').text = weatherTomorrow["precipMM"];
  query('#windDirectionDegree3').text =  '${weatherTomorrow["winddirDegree"]}º';
  query('#windDirection16Point3').text =  weatherTomorrow["winddir16Point"];
  query('#windSpeedkmh3').text =  '${weatherTomorrow["windspeedKmph"]}Kmph';
  query('#windSpeedMiles3').text =  '${weatherTomorrow["windspeedMiles"]}Mph';
  ImageElement image3 = query('#img3');
  image3.src = weatherTomorrow["weatherIconUrl"][0]["value"];
}

File: client-js_&_jquery/tests/weather.js
    function login() {
        var user = document.getElementById("user_id").value;
        var pass = document.getElementById("pass_id").value;

        var req = new XMLHttpRequest();
        req.open("POST", "http://192.168.1.63:8000/weather/login/", true);
        req.setRequestHeader('Content-Type', 'application/json');
        req.withCredentials = true;
        req.onreadystatechange = function() {
            if (req.readyState == XMLHttpRequest.DONE) {
                if (req.status == 200) {        
                    document.getElementById("log_form").style.display = 'none';
                    document.getElementById("logged_user").style.display = 'block';
                    document.getElementById("logged_user").textContent = document.getElementById("user_id").value;
                    document.getElementById("logout_button").style.display = 'block';
                    hide_error();
                } 
                else if (req.status == 401) {
                    document.getElementById('error_text').textContent = "User/password is incorrect";
                    document.getElementById('error').style.display="";
                }
            }
        }
        req.send(JSON.stringify({username: user, password: pass}));
    }

    function logout() {
        var req = new XMLHttpRequest();
        req.open("POST", "http://192.168.1.63:8000/weather/logout/", true);
        req.withCredentials = true;
        req.send();

        document.getElementById('log_form').style.display = '';
        document.getElementById('logged_user').style.display = 'none';
        document.getElementById('logout_button').style.display = 'none';
        document.getElementById('content').style.display = 'none';
        hide_error();
    }

    function newQuery() {
        var city = document.getElementById("city").value;

        var req = new XMLHttpRequest();
        req.open("POST", "http://192.168.1.63:8000/weather/querys/", true);
        req.withCredentials = true;
        req.onreadystatechange = function() {
            if (req.readyState == XMLHttpRequest.DONE) {
                if (req.status == 200) {
                    var j = JSON.parse(req.responseText);
                    if (j.hasOwnProperty("result_json")){
                        hide_error();
                        onSuccess(j["result_json"]);
                    } 
                    else {
                    document.getElementById('error_text').textContent = "City not found";
                    document.getElementById('error').style.display = '';
                    }
                } 
                else if (req.status == 401) {
                    document.getElementById('error_text').textContent = "User not logged in";
                    document.getElementById('error').style.display = '';
                }
            }   
        };
        req.setRequestHeader('Content-Type', 'application/json');
        req.send(JSON.stringify({city: city}));
    }

    function onSuccess(j) {
        document.getElementById('content').style.display = '';
        document.getElementById('city_name').textContent = j["data"]["request"][0]["query"];

        // Current condition
        var currentConditions = j["data"]["current_condition"][0];
        document.getElementById('observationTime').textContent = currentConditions["observation_time"];
        document.getElementById('description1').textContent = currentConditions["weatherDesc"][0]["value"];
        document.getElementById('tCelsius').textContent = currentConditions["temp_C"] + 'ºC';
        document.getElementById('tFarenheit').textContent = currentConditions["temp_F"]+ 'ºF';
        document.getElementById('humidity').textContent = currentConditions["humidity"]+ '%';
        document.getElementById('precipitation1').textContent = currentConditions["precipMM"]+ ' MM';
        document.getElementById('cloudcover').textContent =  currentConditions["cloudcover"]+ '%';
        document.getElementById('visibility').textContent =  currentConditions["visibility"]+ 'Km';
        document.getElementById('pressure').textContent =  currentConditions["pressure"];
        document.getElementById('windDirectionDegree1').textContent =  currentConditions["winddirDegree"]+ 'º';
        document.getElementById('windDirection16Point1').textContent =  currentConditions["winddir16Point"];
        document.getElementById('windSpeedkmh1').textContent =  currentConditions["windspeedKmph"]+ 'Kmph';
        document.getElementById('windSpeedMiles1').textContent =  currentConditions["windspeedMiles"]+ 'Mph';
        document.getElementById('img1').src = currentConditions["weatherIconUrl"][0]["value"];

        // Today's forecast
        var weatherToday = j["data"]["weather"][0];
        document.getElementById('dateToday').textContent = weatherToday["date"];
        document.getElementById('description2').textContent = weatherToday["weatherDesc"][0]["value"];
        document.getElementById('minTCelsius1').textContent = weatherToday["tempMinC"]+ 'ºC';
        document.getElementById('minTFarenheit1').textContent = weatherToday["tempMinF"]+ 'ºF';
        document.getElementById('maxTCelsius1').textContent = weatherToday["tempMaxC"]+ 'ºC';
        document.getElementById('maxTFarenheit1').textContent = weatherToday["tempMaxF"]+ 'ºF';
        document.getElementById('precipitation2').textContent = weatherToday["precipMM"]+ 'MM';
        document.getElementById('windDirectionDegree2').textContent =  weatherToday["winddirDegree"]+ 'º';
        document.getElementById('windDirection16Point2').textContent =  weatherToday["winddir16Point"];
        document.getElementById('windSpeedkmh2').textContent =  weatherToday["windspeedKmph"]+ 'Kmph';
        document.getElementById('windSpeedMiles2').textContent =  weatherToday["windspeedMiles"]+ 'Mph';
        document.getElementById('img2').src = weatherToday["weatherIconUrl"][0]["value"];

        // Tomorrow's forecast
        var weatherTomorrow = j["data"]["weather"][1];
        document.getElementById('dateTomorrow').textContent = weatherTomorrow["date"];
        document.getElementById('description3').textContent = weatherTomorrow["weatherDesc"][0]["value"];
        document.getElementById('minTCelsius2').textContent = weatherTomorrow["tempMinC"]+ 'ºC';
        document.getElementById('minTFarenheit2').textContent = weatherTomorrow["tempMinF"]+ 'ºF';
        document.getElementById('maxTCelsius2').textContent = weatherTomorrow["tempMaxC"]+ 'ºC';
        document.getElementById('maxTFarenheit2').textContent = weatherTomorrow["tempMaxF"]+ 'ºF';
        document.getElementById('precipitation3').textContent = weatherTomorrow["precipMM"]+ 'MM';
        document.getElementById('windDirectionDegree3').textContent =  weatherTomorrow["winddirDegree"]+ 'º';
        document.getElementById('windDirection16Point3').textContent =  weatherTomorrow["winddir16Point"];
        document.getElementById('windSpeedkmh3').textContent =  weatherTomorrow["windspeedKmph"]+ 'Kmph';
        document.getElementById('windSpeedMiles3').textContent =  weatherTomorrow["windspeedMiles"]+ 'Mph';
        document.getElementById('img3').src = weatherTomorrow["weatherIconUrl"][0]["value"];
    }

    function hide_error() {
        document.getElementById('error').style.display = 'none';
    }

    document.getElementById("form_login").addEventListener('submit',  function (eventObject) { eventObject.preventDefault(); login(); }, false);
    document.getElementById("formulary").addEventListener('submit',  function (eventObject) { newQuery(); eventObject.preventDefault(); }, false);
    document.getElementById("close").addEventListener('click',  function (eventObject) { hide_error();}, false);
    document.getElementById("logout").addEventListener('click',  function (eventObject) { logout();}, false);
    
    
//************************************TESTING************************************

    function test() {
        var jsonString1 = '{"data": {"current_condition": [{"windspeedMiles": "4", "winddirDegree": "50", "observation_time": "08:08 AM", "temp_F": "64", "temp_C": "18", "humidity": "68", "winddir16Point": "NE", "pressure": "1019", "precipMM": "0.0", "visibility": "10", "cloudcover": "0", "weatherCode": "113", "weatherDesc": [{"value": "Sunny"}], "windspeedKmph": "7", "weatherIconUrl": [{"value": "http://www.worldweatheronline.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"}]}], "weather": [{"windspeedMiles": "9", "tempMaxC": "31", "tempMaxF": "87", "tempMinC": "17", "tempMinF": "63", "winddir16Point": "E", "weatherCode": "113", "precipMM": "0.0", "weatherDesc": [{"value": "Sunny"}], "winddirDegree": "90", "winddirection": "E", "date": "2012-09-05", "windspeedKmph": "15", "weatherIconUrl": [{"value": "http://www.worldweatheronline.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"}]}, {"windspeedMiles": "9", "tempMaxC": "31", "tempMaxF": "87", "tempMinC": "17", "tempMinF": "63", "winddir16Point": "SE", "weatherCode": "113", "precipMM": "0.0", "weatherDesc": [{"value": "Sunny"}], "winddirDegree": "129", "winddirection": "SE", "date": "2012-09-06", "windspeedKmph": "14", "weatherIconUrl": [{"value": "http://www.worldweatheronline.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"}]}], "request": [{"query": "Madrid, Spain", "type": "City"}]}}';
        var jsonString2 = '{"data": {"current_condition": [{"windspeedMiles": "7", "winddirDegree": "110", "observation_time": "08:18 AM", "temp_F": "77", "temp_C": "25", "humidity": "41", "winddir16Point": "ESE", "pressure": "1017", "precipMM": "0.0", "visibility": "10", "cloudcover": "0", "weatherCode": "113", "weatherDesc": [{"value": "Sunny"}], "windspeedKmph": "11", "weatherIconUrl": [{"value": "http://www.worldweatheronline.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"}]}], "weather": [{"windspeedMiles": "6", "tempMaxC": "28", "tempMaxF": "83", "tempMinC": "15", "tempMinF": "59", "winddir16Point": "ESE", "weatherCode": "113", "precipMM": "0.0", "weatherDesc": [{"value": "Sunny"}], "winddirDegree": "102", "winddirection": "ESE", "date": "2012-09-05", "windspeedKmph": "10", "weatherIconUrl": [{"value": "http://www.worldweatheronline.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"}]}, {"windspeedMiles": "5", "tempMaxC": "29", "tempMaxF": "84", "tempMinC": "16", "tempMinF": "61", "winddir16Point": "SE", "weatherCode": "113", "precipMM": "0.0", "weatherDesc": [{"value": "Sunny"}], "winddirDegree": "132", "winddirection": "SE", "date": "2012-09-06", "windspeedKmph": "8", "weatherIconUrl": [{"value": "http://www.worldweatheronline.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"}]}], "request": [{"query": "Bucharest, Romania", "type": "City"}]}}';
        var json1 = JSON.parse(jsonString1);
        var json2 = JSON.parse(jsonString2);
        testView(json1, json2, 10000);
        testJson(jsonString1, 10000);
        testXHR(1000);
    }
    
    function testXHR(n) {
        testLogin(n, "admin", "admin");
    }
    
    function testLogin(n, user, pass) {
    
        var json = {
            username : user,
            password : pass
        };

        var URL = "http://192.168.1.63:8000/weather/login/";
		
        console.log('Executing ' + n + ' calls to login...');
        var l = 0;
        var start = new Date().getTime();
		for (var i=0; i<n; i++){
		    var req = new XMLHttpRequest();
            req.open("POST", "http://192.168.1.63:8000/weather/login/", true);
            req.setRequestHeader('Content-Type', 'application/json');
            req.withCredentials = true;
            req.onreadystatechange = function() {
                if (this.readyState == XMLHttpRequest.DONE) {
                    l++;
                    if (l == n){
                        var end = new Date().getTime();
                        var elapsed = end - start;
                        var average = elapsed / n;
                        console.log('Average time for ' + n + ' logins = ' + average + ' ms');
                        testQuery(n, "Budapest");
                    }
                }
            };
            req.send(JSON.stringify(json));
        }
    }
    
    function testQuery(n, city) {
	    
        var json = {
            city : city
        };
		
		var URL = "http://192.168.1.63:8000/weather/querys/";
		
		console.log('Executing ' + n + ' calls to newQuery...');
        var q = 0;
        var start = new Date().getTime();

        for (var i=0; i<n; i++){
		    var req = new XMLHttpRequest();
            req.open("POST", "http://192.168.1.63:8000/weather/querys/", true);
            req.withCredentials = true;
            req.onreadystatechange = function() {
                if (this.readyState == XMLHttpRequest.DONE) {
                    q++;
                    if (q == n){
                        var end = new Date().getTime();
                        var elapsed = end - start;
                        var average = elapsed / n;
                        console.log('Average time for ' + n + ' querys = ' + average + ' ms');
                        testQueryView(n, "Yokohama", "Madrid");
                    }
                }
		    };
		    req.setRequestHeader('Content-Type', 'application/json');
            req.send(JSON.stringify(json));
		}
    } 
    
    function testQueryView(n, city1, city2) {
	    
	    var json;
        var json1 = {
            city : city1
        };
        var json2 = {
            city : city2
        };
		
		var URL = "http://192.168.1.63:8000/weather/querys/";
		
		console.log('Executing ' + n + ' calls to newQuery with view update...');
        var v = 0;
        var start = new Date().getTime();
        for (var i=0; i<n; i++){
            if (i % 2 == 0){
                json = json1;
            }
            else {
                json = json2;
            }
            var req = new XMLHttpRequest();
            req.open("POST", "http://192.168.1.63:8000/weather/querys/", true);
            req.withCredentials = true;
            req.onreadystatechange = function() {
                if (this.readyState == XMLHttpRequest.DONE) {
                    if (this.status == 200) {
                        var j = JSON.parse(this.responseText);
                        if (j.hasOwnProperty("result_json")){
                            hide_error();
                            onSuccess(j["result_json"]);
                        } 
                        else {
                        document.getElementById('error_text').textContent = "City not found";
                        document.getElementById('error').style.display = '';
                        }
                    }
                    v++;
                    if (v == n){
                        var end = new Date().getTime();
                        var elapsed = end - start;
                        var average = elapsed / n;
                        console.log('Average time for ' + n + ' querys with view update = ' + average + ' ms');
                    }
                }
		    };
		    req.setRequestHeader('Content-Type', 'application/json');
            req.send(JSON.stringify(json));
        }
    } 
    
    function testView(json1,json2, n){
        var start = new Date().getTime();
        console.log('Executing ' + n + ' view switches...');
        for(var i = 0; i < n; i++){
            onSuccess(json1);
            onSuccess(json2);
        }
        var end = new Date().getTime();
        var elapsed = end - start;
        var average = elapsed / n;
        console.log('Average time for ' + n + ' view switches = ' + average + ' ms');
    }
    
    function testJson(jsonString1, n){
        var start = new Date().getTime();
        console.log('Executing ' + n + ' json parses...');
        for(var i = 0; i < n; i++){
            JSON.parse(jsonString1);
        }
        var end = new Date().getTime();
        var elapsed = end - start;
        var average = elapsed / n;
        console.log('Average time for ' + n + ' json parses = ' + average + ' ms');
    }
    
    // Test event handler
    window.addEventListener("keypress",  function (event) {
        if (event.which == 13) {
            test();
        }
    }, false);

